### PR TITLE
Support custom encoders for primitive types

### PR DESCRIPTION
**Pull request by pkoenig10**
## Before this PR
Using a custom encoder with a primitive types fails with `NoSuchElementException`:
```
at java.base/java.util.Optional.orElseThrow(Optional.java:375)
  	at com.palantir.dialogue.annotations.processor.generate.ServiceImplementationGenerator.underlyingCustomType(ServiceImplementationGenerator.java:200)
  	at com.palantir.dialogue.annotations.processor.generate.ServiceImplementationGenerator.encoder(ServiceImplementationGenerator.java:188)
  	at com.palantir.dialogue.annotations.processor.generate.ServiceImplementationGenerator.lambda$generate$1(ServiceImplementationGenerator.java:78)
  	at java.base/java.util.Optional.map(Optional.java:258)
  	at com.palantir.dialogue.annotations.processor.generate.ServiceImplementationGenerator.lambda$generate$2(ServiceImplementationGenerator.java:78)
  	at com.palantir.dialogue.annotations.processor.data.ParameterTypes$LambdaCases.header(ParameterTypes.java:228)
  	at com.palantir.dialogue.annotations.processor.data.ParameterTypes$Header.match(ParameterTypes.java:280)
  	at com.palantir.dialogue.annotations.processor.data.ParameterTypes$CaseOfMatchers$PartialMatcher.otherwise(ParameterTypes.java:739)
  	at com.palantir.dialogue.annotations.processor.data.ParameterTypes$CaseOfMatchers$PartialMatcher.otherwise_(ParameterTypes.java:743)
  	at com.palantir.dialogue.annotations.processor.generate.ServiceImplementationGenerator.lambda$generate$7(ServiceImplementationGenerator.java:81)
  	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:271)
  	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
  	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
  	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
  	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
  	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
  	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
  	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
  	at com.palantir.dialogue.annotations.processor.generate.ServiceImplementationGenerator.lambda$generate$8(ServiceImplementationGenerator.java:83)
  	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:406)
  	at com.palantir.dialogue.annotations.processor.generate.ServiceImplementationGenerator.generate(ServiceImplementationGenerator.java:73)
```

For context, I'm using Dialogue annotations to define a service interface for testing and want to define something like:
```java
@Request(method = HttpMethod.GET, path = "/foo")
void foo(
    @Header(value = "Cookie", encoder = CookieTokenEncoder.class) BearerToken token,
```
```java
final class CookieTokenEncoder implements ListParamEncoder<BearerToken> {
    @Override
    public List<String> toParamValues(BearerToken token) {
        return List.of("TOKEN=" + token);
    }
}
```

I could wrap it in a custom class, but there doesn't seem to be much benefit to forcing consumers to do that.

## After this PR
Custom encoders can be used with primitive types.

## Possible downsides?
Is there a reason we explicitly did not want to support this? Is there a benefit to only allowing one serialization format for primitive types, even when manually using the Dialogue annotations?

### CODE REVIEW

File: dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java
Comment (pkoenig10) [inline dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java:316]: Is there a reason `generatePlainSerializer` returns an optional? The only empty return value is here and it seems like it would be better to throw if a parameter encoder is not defined for a custom type.

Comment (jkozlowski) [reply dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java:316]: Yes, so that I could build this incrementally, and ignore stuff that wasn't handled :) I think at this point the Optional could be removed.

File: dialogue-annotations-processor/src/test/java/com/palantir/myservice/service/MyCustomStringParameterEncoder.java
Comment (jkozlowski) [inline dialogue-annotations-processor/src/test/java/com/palantir/myservice/service/MyCustomStringParameterEncoder.java:24]: So my first gut reaction is that we should not do this, the wrapping for a custom type costs basically nothing. Certainly for Strings it just feels wrong.

Maybe for something like BearerToken that might make sense, but maybe that just means it's not a primitive/out of the box type?

**Comment by jkozlowski** [comment]
@pkoenig10 perhaps with your example, actual out of the box support for cookies is what we want?

**Comment by pkoenig10** [comment]
Maybe? Given that cookies are just headers though, I'm not sure we should special-case them in that way.

If we do add support cookies, I think we'd want to keep the cookie -> header conversation code inside the annotation processor.

**Comment by pkoenig10** [comment]
But there's probably not a valid use case for this outside of cookies, so adding cookie support is probably the right approach.

I'll put up a PR to see what this would look like.

**Comment by jkozlowski** [comment]
Thanks! My knowledge of cookies is rather rusty, circa 2000, when trying to write some web apps with php. So not clear if that is a good idea or not, but if the serialization format is fixed for cookies, I think out of the box support might be nice.

**Comment by carterkozak** [comment]
I had thought we supported cookie auth in generated dialogue clients, but a quick grep doesn't look like it was ever implemented.
Custom serializers for conjure primitives sound reasonable to me, I could see a string->string case in which we normalize values in some way (`.toUpperCase(Locale.ENGLISH)`).
First-class cookie support sounds great as well, but I'd prefer to start by adding support for cookie auth based on conjure, then implementing it here to avoid duplication.

**Comment by carterkozak** [comment]
Actually, we do support cookie auth out of the box, see https://github.com/palantir/conjure-java/blob/9d0e994aaf5b923bd264923f21ed3b14b523757f/conjure-java-core/src/integrationInput/java/com/palantir/product/CookieServiceBlocking.java#L45

**Comment by pkoenig10** [comment]
@carterkozak Do you have a preference between cookie params and custom encoder?

Given that was already support cookie params in Conjure, that seems like the more appropriate approach here. Because we don't, for example, let you define custom encoders in Conjure.

**Comment by jkozlowski** [comment]
> Custom serializers for conjure primitives sound reasonable to me, I could see a string->string case in which we normalize values in some way (.toUpperCase(Locale.ENGLISH)).

See that's where I feel like a custom alias is a better solution, you're doing some sort of operation on the Strings, so it should probably be specifically identified in your code. Also, why would you need to uppercase your strings?

**Comment by carterkozak** [comment]
> Given that was already support cookie params in Conjure, that seems like the more appropriate approach here. Because we don't, for example, let you define custom encoders in Conjure.

I think we only support cookie auth BearerToken values in conjure, but they use `Cookie: <NAME>=<BEARER_TOKEN>`. Cookies can get difficult if we want to support multiple cookies at once because they must be merged into a single header value.

>See that's where I feel like a custom alias is a better solution, you're doing some sort of operation on the Strings

I'm not sure I agree with this for the annotation processor, the point is not to lock users into the conjure opinions -- we shouldn't force types to be wrapped in order to provide custom encoding, that may expose implementation details to those who consume the client (not necessarily the same folks who wrote the client).

> so it should probably be specifically identified in your code.

Isn't this what the annotations encoder value does?

> Also, why would you need to uppercase your strings?

A contrived example, but it's not so much different from the auth cookie bearertoken.

**Comment by carterkozak** [comment]
(accidental close -- reopened)

**Comment by pkoenig10** [comment]
> Cookies can get difficult if we want to support multiple cookies at once because they must be merged into a single header value.

I was planning to only support a single value. I don't think there are many legitimate use cases for sending multiple cookie values and limiting to a single value better aligns with the cookie support in Conjure.

Unless you think there's a good chance we'd regret that limitation?

**Comment by carterkozak** [comment]
> Do you have a preference between cookie params and custom encoder?

I think we want both, custom encoders are likely easiest to do first and unblock.

**Comment by jkozlowski** [comment]
@pkoenig10 for posterity, can you link where you need this for btw?

**Review by jkozlowski — APPROVED** [review]

**Comment by carterkozak** [comment]
> Unless you think there's a good chance we'd regret that limitation?

I imagine we may end up with both cookie auth and a second cookie header, no?

**Comment by pkoenig10** [comment]
Yeah I realized this doesn't quite work. You either need to limit endpoints to 1 cookie parameter, which feels a bit limiting, or add the logic to merge cookies, which is fairly complex. Going to punt this for now.

**Pull request merged**
